source/language/openpulse.rst

```diff
@@ -30,227 +31,249 @@ could even reference multiple qubits in a chain.
-of qubits does not matter.
-
-Frames are retrieved using the ``frameof`` function and passing in the name and
-list of physical qubits.
-
-.. code-block: none
-
-   frameof("drive", %0)
-   
-   // These next two lines refer to the same frame
-   frameof("coupling", %0, %1)
-   frameof("coupling", %1, %0)
-
-Frame names may seem like they ascribe meaning or that there are only certain
-permissable names. This is not the case; frame names are arbitrary. The frames
-are later mapped to channel names that do have meaning for a certain hardware
-vendor. For example, the hardware vendor may choose to map frames to channels
-using JSON:
-
-.. code-block: javascript
-
-   {
-     drive: {
-       "{0}": "channel0",
-       "{1}": "channel1"
-     coupling: {
-       "{0,1}": "channel2"
-     }
-   }
+When interacting with qubits, it turns out to be quite useful to keep track of a frame of reference,
+akin to the rotating frame of a Hamiltonian, throughout the execution of a program. These ``frame``
+types are responsible for two things
+
+- Tracking time appropriately so programs do not need to deal in absolute time or with the
+  bookkeeping of advancing time in a sequence of pulses.
+- Tracking phase by producing a complex value given an input time (i.e. via the mathematical
+  relationship :math:`e^{i\left(2\pi f t + \theta\right)}`,  where `f` is frequency and
+  :math:`\theta` is phase). One motivation for keeping track of phase is to allow pulses to be
+  defined in the rotating frame (i.e. without the carrier) with the effect being
+  an equivalent application in the lab frame (i.e. with the carrier supplied by the ``frame``).
+  Another motivation is to more naturally implement a "virtual Z-gate", which does not require a
+  physical pulse but rather shifts the phase of all future pulses on that frame.
+
+The frame is composed of three parts:
+
+1. A frequency ``frequency`` of type ``float``.
+2. A phase ``phase`` of type ``angle``.
+3. A time which is manipulated implicitly and cannot be modified other
+   than through existing timing instructions like ``delay`` and ``barrier``. The time increment
+   is determined by the channel on which the frame is played (see :ref:`Timing` section).
+
+The ``frame`` type is a virtual resource and the exact precision of these parameters is
+hardware specific. It is thus up to the compiler to choose how to implement the required
+transformations to physical resources in hardware (e.g. mapping multiple frames to a
+single NCO).
+
+Frame Construction
+~~~~~~~~~~~~~~~~~~
 
-This has the advantage that one can run any program with any arbitrary frame
-names provided a mapping to the appropriate channels is given.
+Frames are purely virtual and can be constructed using the ``Frame`` command
 
-Restrictions on the use of frames
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+.. code-block:: javascript
 
-There are two important considersations when dealing with frames.
+  frame driveframe = Frame(5e9, 0.0);
```

Comment:
Can we add the function signature, ie `Frame(float[size] frequency, angle[size] phase)`?